README.md
# Test Helper

This project is inspired by Python minitest's some useful methods which could be used in test.
The below are useful functions:

    p, pp, pl, ppl

github: [https://github.com/jichen3000/node_testhelper](https://github.com/jichen3000/node_testhelper)

-----------------------

### Author

Colin Ji <jichen3000@gmail.com>


### How to install

    npm install testhelper

### How to use

p, print with title. This function will print variable name as the title.
<br>code:
    
    obj = {foo:"bar", baz: {aa:1, bb:2}};
    // add a title 'obj' automatically.
    obj.p()

print result:

    obj : { foo: 'bar', baz: { aa: 1, bb: 2 } }

pl, print with title and code loction. This function just like pt, but will print
the code location at the first line.
And some editors support to go to the line of that file, such as Sublime2.
Notice: it will print a null line before the location information.
<br>code:
    
    obj = {foo:"bar", baz: {aa:1, bb:2}};
    // add a empty line, and a code position line.
    // add a title 'obj' automatically.
    obj.pl()

print result:


      File "/Users/colin/work/node_testhelper/lib/testhelper.js", line 21, in <anonymous>
    obj : { foo: 'bar', baz: { aa: 1, bb: 2 } }

pp, prety print with title. This function will print variable name as the title.
<br>code:
    
    obj = {foo:"bar", baz: {aa:1, bb:2}};
    // add a title 'obj' automatically.
    obj.pp()

print result:

    obj : 
    {
      "foo": "bar",
      "baz": {
        "aa": 1,
        "bb": 2
      }
    }

ppl, pretty print with title and code loction. This function just like pt, but will print
the code location at the first line.
And some editors support to go to the line of that file, such as Sublime2.
Notice: it will print a null line before the location information.
<br>code:
    
    obj = {foo:"bar", baz: {aa:1, bb:2}};
    // add a empty line, and a code position line.
    // add a title 'obj' automatically.
    obj.ppl()

print result:


      File "/Users/colin/work/node_testhelper/lib/testhelper.js", line 52, in <anonymous>
    obj : 
    {
      "foo": "bar",
      "baz": {
        "aa": 1,
        "bb": 2
      }
    }

Notice, for the number, please use the functions like the below:

    (123).p()

and, these functions are not numerable for the object, which means:

    var a_obj = {bb:1};
    for (key in a_obj) console.log(key);

would not print the 'p','pl' these functions in the object.


CHANGES.md
### 1.5.3 (2014-12-17)

- Correct print array, and refactor, now separate the commons.js.

### 1.5.2 (2014-12-12)

- Correct print NaN in pp, and add assert all output.

### 1.5.1 (2014-12-12)

- Print string normally.

### 1.5.0 (2014-12-11)

- Support print true and false in p function.

### 1.4.0 (2014-12-11)

- Support print NaN as 'NaN'.

### 1.3.1 (2014-12-10)

- p, pl functions support print number value.

### 1.2.1 (2014-12-10)

- Let p, pl, pp, ppl functions are not enumerable for object.

### 1.1.1 (2014-12-10)

- Print function content as a string, not [Function]

### 1.0.1 (2014-12-08)

- Correct a typo in README.md.

### 1.0.0 (2014-12-08)

- Add `pp` and `ppl` method.

### 0.0.3 (2014-12-07)

- Add `pl` method.

### 0.0.2 (2014-12-07)

- Change the configure file.

### 0.0.1 (2014-12-07)

- Add `p` method.


test/test_publish.js
require('testhelper');

if (require.main === module) {
  obj = {foo:"bar", baz: {aa:1, bb:2}};
  p(obj); // ignore this situation, return null
  obj.p(); // obj : { foo: 'bar', baz: { aa: 1, bb: 2 } }, return obj    
  obj.pl();
  obj.pp();
  obj.ppl();
}

test/test_testhelper.js
require('../lib/testhelper');

if (require.main === module) {
  var obj = {foo:"bar", baz: {aa:1, bb:2}};
  p(obj); // ignore this situation, return null
  obj.p(); // obj : { foo: 'bar', baz: { aa: 1, bb: 2 } }, return obj

  obj.pl();

  obj.pp();
  
  obj.ppl();
  // pp(obj);

  var a = {bb:1}
  for (key in a) console.log(key);

  // number is not yet supported
  (1).p();
  (typeof 1).pp();
  "123".p();
  true.pp();
  [1,2].p();
  [[1,2]].pp();

  var the_func = function (argument) {
      argument.pp();
  };

  the_func.pp();
  // null.pp();
  var the_null = null;
  // the_null.pp();
}

lib/stack_info.js
var fs = require('fs');
var util = require('util');

var stackInfo = module.exports;

function getStack () {
  // callsite
  var orig = Error.prepareStackTrace;
  Error.prepareStackTrace = function(_, stack) {
    return stack;
  };
  var err = new Error;
  Error.captureStackTrace(err, arguments.callee);
  var stack = err.stack;
  Error.prepareStackTrace = orig;
  return stack;
}

stackInfo.getCalleeStrSync = function (filePath, lineNumber, columnNumber) {
  var data = fs.readFileSync(filePath, {encoding: 'utf-8'});
  var curLine = data.split(/\r?\n/)[lineNumber-1];
  var callee = curLine.substring(0, columnNumber);
  callee = callee.trim()
  callee = callee.substring(0, callee.lastIndexOf("."));
  return callee;
}

stackInfo.getCodePosition = function (filePath, lineNumber, columnNumber, funcName) {
  //   File "/Users/colin/work/node_testhelper/lib/testhelper.js", line 30, in <anonymous>
  return util.format('File "%s", line %d, in %s', filePath, lineNumber, funcName);
}

stackInfo.getPositionInfo = function (frameIndex){
  // frameIndex's default value = 2
  frameIndex = typeof frameIndex !== 'undefined' ? frameIndex : 2;
  var curFrame = getStack()[frameIndex];
  return [curFrame.getFileName(), curFrame.getLineNumber(), 
      curFrame.getColumnNumber(), curFrame.getFunctionName() || '<anonymous>']
};
stackInfo.getCurrentFramePositionInfo = function (frameIndex){
  frameIndex = 1;
  var curFrame = getStack()[frameIndex];
  return [curFrame.getFileName(), curFrame.getLineNumber(), 
      curFrame.getColumnNumber(), curFrame.getFunctionName() || '<anonymous>']
};
stackInfo.getCallerFramePositionInfo = function (frameIndex){
  frameIndex = 2;
  var curFrame = getStack()[frameIndex];
  return [curFrame.getFileName(), curFrame.getLineNumber(), 
      curFrame.getColumnNumber(), curFrame.getFunctionName() || '<anonymous>']
};

if (require.main === module){
  var assert = require('assert');
  var puts = console.log;
  var posInfo = stackInfo.getPositionInfo(1);
  // assert.deepEqual(posInfo, ["/Users/colin/work/node_testhelper/lib/stack_info.js",40,25,"anonymous"]);
  assert.deepEqual(posInfo[0], "/Users/colin/work/node_testhelper/lib/stack_info.js");

  // posInfo = [ '/Users/colin/work/node_testhelper/lib/stack_info.js',
  //             56,
  //             27,
  //             '<anonymous>' ]
  // puts(posInfo);

  assert.deepEqual(stackInfo.getCalleeStrSync.apply(this, posInfo), 'var posInfo = stackInfo');

  assert.equal(stackInfo.getCodePosition("/Users/colin/work/node_testhelper/lib/stack_info.js",46,25,"<anonymous>"),
      'File "/Users/colin/work/node_testhelper/lib/stack_info.js", line 46, in <anonymous>');
}

lib/testhelper.js
var prettyPrintString = require("./pretty_print_string");
var stackInfo = require("./stack_info");
var originalThis = require("./original_this");
var util    = require("util");
var puts = console.log;


var p = function () {
  var posInfo = stackInfo.getCallerFramePositionInfo();
  // puts(posInfo);
  calleeStr = stackInfo.getCalleeStrSync.apply(null, posInfo);
  if (calleeStr===''){return null;}
  var originalObject = originalThis(this);
  puts(util.format('%s : %s', calleeStr, prettyPrintString.genSimpleStr(originalObject)));
  return originalObject;
};

var pl = function () {
  var posInfo = stackInfo.getCallerFramePositionInfo();
  calleeStr = stackInfo.getCalleeStrSync.apply(null, posInfo);
  if (calleeStr===''){return null;}
  var originalObject = originalThis(this);
  puts("");
  puts("  "+stackInfo.getCodePosition.apply(null, posInfo));
  puts(util.format('%s : %s', calleeStr, prettyPrintString.genSimpleStr(originalObject)));
  return originalObject;
};

var pp = function () {
  var posInfo = stackInfo.getCallerFramePositionInfo();
  calleeStr = stackInfo.getCalleeStrSync.apply(null, posInfo);
  if (calleeStr===''){return null;}
  var originalObject = originalThis(this);
  puts(util.format('%s : \n%s', calleeStr, prettyPrintString.genPrettyStr(originalObject)));
  return originalObject;
};

var ppl = function () {
  var posInfo = stackInfo.getCallerFramePositionInfo();
  calleeStr = stackInfo.getCalleeStrSync.apply(null, posInfo);
  if (calleeStr===''){return null;}
  var originalObject = originalThis(this);
  puts("");
  puts("  "+stackInfo.getCodePosition.apply(null, posInfo));
  puts(util.format('%s : \n%s', calleeStr, prettyPrintString.genPrettyStr(originalObject)));
  return originalObject;
};

Object.defineProperty( Object.prototype, "p", {
  value: p,
  enumerable: false
});

Object.defineProperty( Object.prototype, "pl", {
  value: pl,
  enumerable: false
});

Object.defineProperty( Object.prototype, "pp", {
  value: pp,
  enumerable: false
});

Object.defineProperty( Object.prototype, "ppl", {
  value: ppl,
  enumerable: false
});

Object.defineProperty( Number.prototype, "p", {
  value: p,
  enumerable: false
});

Object.defineProperty( Number.prototype, "pl", {
  value: pl,
  enumerable: false
});

Object.defineProperty( Number.prototype, "pp", {
  value: pp,
  enumerable: false
});

Object.defineProperty( Number.prototype, "ppl", {
  value: ppl,
  enumerable: false
});

if ( require.main === module) {
  var assert = require('assert');
  // for test log messages
  var msgCache = [];

  function getMsgs (isClean) {
    isClean = typeof isClean !== 'undefined' ? isClean : true;
    var msgArr = msgCache.slice(0);
    if (isClean) msgCache.length = 0;
    return msgArr;
  }

  var oldLog = puts;
  puts = function (msg) {
    msgCache.push(msg);
    oldLog.apply(console, arguments);
  }
  // end

  var obj = {foo:"bar", baz: {aa:1, bb:2}};
  p(obj); // ignore this situation, return null
  obj.p(); // obj : { foo: 'bar', baz: { aa: 1, bb: 2 } }, return obj
  assert.deepEqual(getMsgs(), ["obj : { foo: 'bar', baz: { aa: 1, bb: 2 } }"]);

  obj.pl();
  assert.deepEqual(getMsgs(), 
      [ '',
        '  File "/Users/colin/work/node_testhelper/lib/testhelper.js", line 113, in <anonymous>',
        'obj : { foo: \'bar\', baz: { aa: 1, bb: 2 } }' ]);

  obj.pp();
  assert.deepEqual(getMsgs(), [ 'obj : \n{\n  "foo": "bar",\n  "baz": {\n    "aa": 1,\n    "bb": 2\n  }\n}' ]);
  
  obj.ppl();
  assert.deepEqual(getMsgs(), 
      [ '',
        '  File "/Users/colin/work/node_testhelper/lib/testhelper.js", line 122, in <anonymous>',
        'obj : \n{\n  "foo": "bar",\n  "baz": {\n    "aa": 1,\n    "bb": 2\n  }\n}' ]);
  pp(obj);

  // notice, the p, pl would not appear in the a_obj object.
  var a_obj = {bb:1};
  for (key in a_obj) puts(key);
  assert.deepEqual(getMsgs(), [ 'bb' ]);

  the_num = 123;
  the_num.p();
  assert.deepEqual(getMsgs(), [ 'the_num : 123' ]);
  the_num.pp();
  assert.deepEqual(getMsgs(), [ 'the_num : \n123' ]);

  (123).p();
  assert.deepEqual(getMsgs(), [ '(123) : 123' ]);

  NaN.p();
  assert.deepEqual(getMsgs(), [ 'NaN : NaN' ]);
  NaN.pp();
  assert.deepEqual(getMsgs(), [ 'NaN : \n"NaN"' ]);

  var obj={b:1, a:NaN}
  obj.p();
  assert.deepEqual(getMsgs(), [ 'obj : { b: 1, a: NaN }' ]);
  obj.pp();
  assert.deepEqual(getMsgs(), [ 'obj : \n{\n  "b": 1,\n  "a": "NaN"\n}' ]);

  true.p();
  assert.deepEqual(getMsgs(), [ 'true : true' ]);
  false.p();
  assert.deepEqual(getMsgs(), [ 'false : false' ]);
  true.pp(); 
  assert.deepEqual(getMsgs(), [ 'true : \ntrue' ]);

  "ok".p(); 
  assert.deepEqual(getMsgs(), [ "\"ok\" : ok" ]);
  "ok".pp(); 
  assert.deepEqual(getMsgs(), [ "\"ok\" : \n\"ok\"" ]);

  [1,2].p();
  assert.deepEqual(getMsgs(), [ '[1,2] : [ 1, 2 ]' ]);
  [1,2].pp();
  assert.deepEqual(getMsgs(), [ '[1,2] : \n[\n  1,\n  2\n]' ]);
  [{a:1},{b:2}].p();
  assert.deepEqual(getMsgs(), [ '[{a:1},{b:2}] : [ { a: 1 }, { b: 2 } ]' ]);
  [{a:1},{b:2}].pp();
  assert.deepEqual(getMsgs(), [ '[{a:1},{b:2}] : \n[\n  {\n    "a": 1\n  },\n  {\n    "b": 2\n  }\n]' ]);
  ['s','t'].p();
  assert.deepEqual(getMsgs(), [ "['s','t'] : [ 's', 't' ]" ]);
  ['s','t'].pp();
  assert.deepEqual(getMsgs(), [ "['s','t'] : \n[\n  \"s\",\n  \"t\"\n]" ]);

}


lib/commons.js
var fs         = require('fs');
var util       = require('util');

var commons = module.exports;
var puts = console.log;

function getStack () {
  // callsite
  var orig = Error.prepareStackTrace;
  Error.prepareStackTrace = function(_, stack) {
    return stack;
  };
  var err = new Error;
  Error.captureStackTrace(err, arguments.callee);
  var stack = err.stack;
  Error.prepareStackTrace = orig;
  return stack;
}

commons.getCalleeStrSync = function (filePath, lineNumber, columnNumber) {
  var data = fs.readFileSync(filePath, {encoding: 'utf-8'});
  var curLine = data.split(/\r?\n/)[lineNumber-1];
  var callee = curLine.substring(0, columnNumber);
  callee = callee.trim()
  callee = callee.substring(0, callee.lastIndexOf("."));
  return callee;
}

commons.getCodePosition = function (filePath, lineNumber, columnNumber, funcName) {
  //   File "/Users/colin/work/node_testhelper/lib/testhelper.js", line 30, in <anonymous>
  return util.format('  File "%s", line %d, in %s', filePath, lineNumber, funcName);
}

commons.getPositionInfo = function (frameIndex){
  // frameIndex's default value = 2
  frameIndex = typeof frameIndex !== 'undefined' ? frameIndex : 2;
  var curFrame = getStack()[frameIndex];
  return [curFrame.getFileName(), curFrame.getLineNumber(), 
      curFrame.getColumnNumber(), curFrame.getFunctionName() || '<anonymous>']
};


function objectToString(o) {
  return Object.prototype.toString.call(o);
}
function isUndefined(arg) {
  return arg === void 0;
}
function isNumberWrong(arg) {
  return typeof arg === 'number';
  // return !isNaN(parseFloat(arg));
}
function isNumber(arg) {
  // return typeof arg === 'number';
  return !isNaN(parseFloat(arg));
}
function isRegExp(re) {
  return isObject(re) && objectToString(re) === '[object RegExp]';
}
function isFunction(arg) {
  return typeof arg === 'function';
}
function isObject(arg) {
  return typeof arg === 'object' && arg !== null;
}
function isNaNObject(arg) {
  return Object.keys(arg).length===0 && isNaN(arg);
}
function isString(arg) {
  return typeof arg === 'string' || arg instanceof String;
}
commons.genSimpleStr = function (obj) {
  if(isString(obj)) return obj;
  // for Boolean
  if(obj.toString() === 'true' || obj.toString() === 'false') return obj.toString();
  // for NaN
  if(isObject(obj) && Object.keys(obj).length===0 && isNaN(obj)) return 'NaN';
  if (isNumber(obj)) return obj.toString();
  return util.inspect(obj);
}

commons.genPrettyStr = function (obj) {
  if(isObject(obj) && Object.keys(obj).length===0 && isNaN(obj)) return 'NaN';
  var cache = [];
  var result = JSON.stringify(obj, function (key, value) {
    if (isObject(value)){
      if (cache.indexOf(value) !== -1) {
        // circular reference found
        return '[circular]';
      }
      cache.push(value);
    }
    // puts("v:"+value);
    if(!isString(value) && isNaN(value) && !isObject(value) && !isFunction(value)) return 'NaN';
    // if (isUndefined(value)) {
    //   return '' + value;
    // }
    if (isNumber(value) && (isNaN(value) || !isFinite(value))) {
      return value.toString();
    }
    if (isFunction(value) || isRegExp(value)) {
      return value.toString();
      // return '[function]';
    }
    return value;
    
  }, 2);
  cache=null;
  return result;
}

if (require.main === module){
  var assert = require('assert');
  var posInfo = commons.getPositionInfo(1);
  // assert.deepEqual(posInfo, ["/Users/colin/work/node_testhelper/lib/commons.js",40,25,"anonymous"]);
  assert.deepEqual(posInfo[0], "/Users/colin/work/node_testhelper/lib/commons.js");

  assert.deepEqual(commons.getCalleeStrSync.apply(this, posInfo), 'var posInfo = commons');

  assert.equal(commons.getCodePosition("/Users/colin/work/node_testhelper/lib/commons.js",46,25,"<anonymous>"),
      '  File "/Users/colin/work/node_testhelper/lib/commons.js", line 46, in <anonymous>');

  var obj = {mykey:1, myfun:function (argument) {
    console.log(argument);
    return argument;
  }};
  obj.obj = obj;

  assert.equal(commons.genSimpleStr(123), '123');
  var returnThis = function (arg) {
    return arg;
  }
  var mm=123;
  assert.equal(isNumber(returnThis(mm)), true);
  assert.equal(commons.genSimpleStr(mm), '123');
  assert.equal(commons.genSimpleStr(true), 'true');
  assert.equal(commons.genSimpleStr(NaN), 'NaN');
  assert.equal(commons.genSimpleStr(returnThis), '[Function]');
  assert.equal(commons.genPrettyStr(returnThis), "\"function (arg) {\\n    return arg;\\n  }\"");

  assert.equal(commons.genPrettyStr(obj), 
    '{\n  "mykey": 1,\n  "myfun": "function (argument) {\\n    console.log(argument);\\n    return argument;\\n  }",\n  "obj": "[circular]"\n}')
}